File: app/src/main/java/com/kabirnayeem99/paymentpaid/adapters/WorkAdapter.java
package com.kabirnayeem99.paymentpaid.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.kabirnayeem99.paymentpaid.R;
import com.kabirnayeem99.paymentpaid.data.db.entities.Work;
import com.kabirnayeem99.paymentpaid.utils.CustomUtils;

import java.util.ArrayList;
import java.util.List;

public class WorkAdapter extends ListAdapter<Work, WorkAdapter.ViewHolder> {
    public static DiffUtil.ItemCallback<Work> DIFF_CALLBACK = new DiffUtil.ItemCallback<Work>() {
        @Override
        public boolean areItemsTheSame(@NonNull Work oldItem, @NonNull Work newItem) {
            return oldItem.getId() == newItem.getId();
        }

        @Override
        public boolean areContentsTheSame(@NonNull Work oldItem, @NonNull Work newItem) {
            return oldItem.getName().equals(newItem.getName());
        }
    };
    private OnClickListener listener;

    public WorkAdapter() {
        super(DIFF_CALLBACK);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        // when there is not view holder, it creates a new view holder,
        // which later will be updated from the bound data.

        // Here the layout inflater takes the work list item layout file
        // and turns it into a view
        // later these views are used to draw on screen
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_work,
                parent, false);

        // a new view holder is created based on the view of work list item.
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        // binds the new data to the view holder without creating new view holder to reduce cpu
        // as well as the memory usage.

        // gets the data from the database, i.e. work title, submission date and payment amount.
        String workListItemTitle = getItem(position).getName();
        String workListItemDate = getItem(position).getDate();
        String workListItemStudentName = getItem(position).getStudentName();
        String workListItemPayment = String.format("%s", getItem(position).getPayment());

        // binds the data for each of the work got from the db to the existing adapter based on the
        // screen time of the lists item.
        holder.tvWorkName.setText(workListItemTitle);
        holder.tvDate.setText(workListItemDate);
        holder.tvStudentName.setText(workListItemStudentName);
        holder.tvPayment.setText(CustomUtils.formatMoney(workListItemPayment));
    }





    public Work getWorkByPosition(int position) {
        return getItem(position);
    }

    public void setOnClickListener(OnClickListener listener) {
        this.listener = listener;
    }

    public interface OnClickListener {
        void onItemClick(Work work);
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        final TextView tvWorkName, tvDate, tvPayment, tvStudentName;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            tvWorkName = itemView.findViewById(R.id.tv_work_name_list_item_work);
            tvDate = itemView.findViewById(R.id.tv_date_list_item_work);
            tvPayment = itemView.findViewById(R.id.tv_payment_amount_list_item_work);
            tvStudentName = itemView.findViewById(R.id.tv_student_name_list_item_work);

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    listener.onItemClick(getItem(getAdapterPosition()));
                }
            });
        }
    }
}

File: app/src/main/java/com/kabirnayeem99/paymentpaid/adapters/WorkAdapter.kt
package com.kabirnayeem99.paymentpaid.adapters

import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import androidx.recyclerview.widget.AsyncListDiffer
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.kabirnayeem99.paymentpaid.R
import com.kabirnayeem99.paymentpaid.data.db.entities.Work
import com.kabirnayeem99.paymentpaid.other.Utils
import kotlinx.android.synthetic.main.list_item_work.view.*
import java.util.*

/**
 * The <h1>WorkRecyclerView</h1> needs this [WorkAdapter]
 * to populate the views in each row with the [Work] data.
 */
class WorkAdapter : RecyclerView.Adapter<WorkAdapter.ViewHolder>(), Filterable {

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

    /*
    Callback for calculating the differences between
    the newly created list and the already existed list.
     */
    private var differCallBack: DiffUtil.ItemCallback<Work> = object : DiffUtil.ItemCallback<Work>() {
        override fun areItemsTheSame(oldItem: Work, newItem: Work): Boolean {
            return oldItem.documentId == newItem.documentId
        }

        override fun areContentsTheSame(oldItem: Work, newItem: Work): Boolean {
            return oldItem == newItem
        }

    }

    /*
    Helper for computing the difference between
    two lists via DiffUtil on a background thread.
     */
    var differ = AsyncListDiffer(this, differCallBack)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): WorkAdapter.ViewHolder {
        return ViewHolder(
                LayoutInflater
                        .from(parent.context)
                        .inflate(R.layout.list_item_work, parent, false)
        )
    }

    // creates a null on item click listener
    private var onItemClickListener: ((Work) -> Unit)? = null

    override fun onBindViewHolder(holder: WorkAdapter.ViewHolder, position: Int) {
        val work = differ.currentList[position]

        val workListItemTitle = work.name
        val workListItemDate = "${work.day}-${Utils.padMonth(work.month)}-${work.year}"
        val workListItemStudentName = work.studentName
        val workListItemPayment: String = work.payment.toString()

        holder.itemView.apply {
            tvWorkNameListItemWork.text = workListItemTitle
            tvStudentNameListItemWork.text = workListItemStudentName
            tvPaymentAmountListItemWork.text = workListItemPayment
            tvDateListItemWork.text = workListItemDate

            // pass the onItemClickListener the exact work object that was pressed on
            setOnClickListener { onItemClickListener?.let { it(work) } }

        }
    }


    override fun getItemCount(): Int {
        return differ.currentList.size

    }

    /**
     * Register a callback to be invoked when the work tile is clicked.
     *
     * @param listener The callback that will run
     */
    fun setOnItemClickListener(listener: (Work) -> Unit) {
        onItemClickListener = listener
    }


    override fun getFilter(): Filter {
        return filter
    }

    private var filter: Filter = object : Filter() {
        /*
         Invoked in a worker thread to filter the data according
         to the constraint. Results computed by the filtering
         operation are returned as a Filter
        */
        override fun performFiltering(searchEntry: CharSequence?): FilterResults {
            val filteredWorks: MutableList<Work> = mutableListOf()


            if (searchEntry == null || searchEntry.isEmpty() || searchEntry.length <= 1) {
                filteredWorks.addAll(differ.currentList)
                Log.d(TAG, "performFiltering: $filteredWorks")
            } else {
                val searchEntryPattern = searchEntry.toString().toLowerCase(Locale.getDefault()).trim()
                for (work in differ.currentList) {
                    if (work.name.toLowerCase(Locale.ROOT)
                                    .contains(searchEntryPattern) || work.studentName.toLowerCase(Locale.getDefault())
                                    .contains(searchEntryPattern)
                    ) {
                        filteredWorks.add(work)
                    }
                }
            }
            val filterResults = FilterResults()
            filterResults.values = filteredWorks
            return filterResults
        }

        override fun publishResults(constraint: CharSequence?, results: FilterResults?) {
            differ.submitList(results?.values as List<Work>)
        }
    }

    companion object {
        private const val TAG = "WorkAdapter"
    }
}